x/rollapp/keeper/state_info.go

```diff
@@ -85,3 +93,53 @@ func (k Keeper) GetAllStateInfo(ctx sdk.Context) (list []types.StateInfo) {
 
 	return
 }
+
+// DeleteStateInfoUntilTimestamp deletes all stateInfo until the given timestamp
+func (k Keeper) DeleteStateInfoUntilTimestamp(ctx sdk.Context, endTimestampExcl time.Time) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.StateInfoKeyPrefix))
+	storeTS := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.TimestampedStateInfoKeyPrefix))
+
+	// Note that for the active sequencer, the latest state info index will not be within the range of
+	// the state updates to be deleted, as it will be more recent.
+	// For a sequencer that is inactive for 21 days or more, it will be within range, furthermore
+	// the latest state info index and the latest finalized state info index will be the same.
+	latestIndexes := k.GetAllLatestStateInfoIndex(ctx)
+	skipStateInfoIndexes := make(map[string]uint64, len(latestIndexes))
+	for _, index := range latestIndexes {
+		skipStateInfoIndexes[index.RollappId] = index.Index
+	}
+
+	k.IterateStateInfoWithTimestamp(storeTS, endTimestampExcl.UnixMicro(), func(keyTS []byte) bool {
+		key := keyTS[types.TimestampPrefixLen:] // remove the timestamp prefix
+		// skip latest stateInfo and latest finalized stateInfo
+		stateInfoIndex := types.StateInfoIndexFromKey(key)
```

Comment:
u can handle the `TimestampPrefixLen` in the `StateInfoIndexFromKey`
(`StateInfoIndexFromTimestampKey`)